### PR TITLE
[WIP] [BPK-969] Enhance Dangerfile checks

### PR DESCRIPTION
**Pull request by shaundon**
Danger will now check that shrinkwrap changes are intentional.
Also Danger will ensure new files are licensed.

+ [x] For any new web components I've made sure that the style can be extended by the consumer using a `className` override.
+ [x] For any new native components I've made sure that the style can be extended by the consumer using a `style` override.
+ [x] For any new files I've included the [Apache 2.0 License header](https://github.com/Skyscanner/backpack/blob/master/packages/bpk-tokens/formatters/license-header.js#L2-L16) at the top of the file.
+ [x] I've updated `changelog.md` for any changes that need to be highlighted in the changelog.
+ [x] If I've updated `npm-shrinkwrap.json` those changes are intentional.

### CODE REVIEW

File: dangerfile.js [hunk 1]
```diff
@@ -23,34 +23,37 @@
 import includes from 'lodash.includes';
 import { danger, warn, message } from 'danger';
 
+const fs = require('fs');
```
Comment (matthewdavidson) [inline dangerfile.js]: use es2015 module syntax:
```
import fs from 'fs';
```

File: dangerfile.js [hunk 2]
```diff
@@ -71,3 +74,21 @@ const snapshotsModified = fileChanges.some(filePath => (
...
+// New files should include the Backpack license heading.
+const unlicensedFiles = createdFiles.filter((filePath) => {
+  if (filePath.match(/\.(js|css|scss|sh)/)) {
+    const fileContent = fs.readFileSync(filePath);
+    return !fileContent.includes('Backpack - Skyscanner\'s Design System');
+  }
+  return false;
+});
+if (unlicensedFiles.length > 0) {
+  warn(`These new files are do not include the license heading: ${unlicensedFiles.join(', ')}`);
```
Comment (matthewdavidson) [inline dangerfile.js]: I think this should fail the build.

File: dangerfile.js [hunk 3]
```diff
@@ -71,3 +74,21 @@ const snapshotsModified = fileChanges.some(filePath => (
...
+const shrinkwrapUpdated = includes(modifiedFiles, 'changelog.md');
+if (shrinkwrapUpdated) {
+  warn('`npm-shrinkwrap.json` was updated. Ensure that this was intentional.');
+}
+
+// New files should include the Backpack license heading.
+const unlicensedFiles = createdFiles.filter((filePath) => {
+  if (filePath.match(/\.(js|css|scss|sh)/)) {
+    const fileContent = fs.readFileSync(filePath);
+    return !fileContent.includes('Backpack - Skyscanner\'s Design System');
```
Comment (matthewdavidson) [inline dangerfile.js]: Hmm nitpick: i'd have chosen something a little more obvious such as `Licensed under the Apache License, Version 2.0 (the "License");`.

File: dangerfile.js (hunk 1 above)
Comment (shaundon) [reply dangerfile.js]: Yep sorry - I was on autopilot!

**Comment by shaundon** [comment]
My previous comment that it won't let me edit should have said "addressed".

**Comment by shaundon** [comment]
Couple of things I'd like to highlight for reviewers:

* The 'encourage smaller PRs' rule has a threshold of 8 files. This was just an arbitrary decision by me, we can change it if the squad decides to.
* We only check for licences in new files that end in .js, .css, .scss and .sh. I'm basing this on a search I did on the codebase to see which file types had licences.

File: dangerfile.js [hunk 4]
```diff
@@ -71,3 +73,27 @@ const snapshotsModified = fileChanges.some(filePath => (
...
+const shrinkwrapUpdated = includes(modifiedFiles, 'changelog.md');
+if (shrinkwrapUpdated) {
+  warn('`npm-shrinkwrap.json` was updated. Ensure that this was intentional.');
+}
+
+// New files should include the Backpack license heading.
+const unlicensedFiles = createdFiles.filter((filePath) => {
+  if (filePath.match(/\.(js|css|scss|sh)/)) {
+    const fileContent = fs.readFileSync(filePath);
+    return !fileContent.includes('Licensed under the Apache License, Version 2.0 (the "License")');
```
Comment (georgegillams) [inline dangerfile.js]: `license-header.js` exports the license for uses like this. Better to check against this if possible (but I don't know if that is possible)

Comment (shaundon) [reply dangerfile.js]: I did think about that, but I have concerns for a couple of reasons:

* I feel like doing `String.includes` on such a large text snippet might be slower than doing it on a small one.
* `dangerfile.js` exports an XML style comment and a block style one. We'd need to introduce logic to `Dangerfile.js` to know which one of these to use in the check based on the file extension. I dunno - feels a bit like overkill to me for a simple check!

What do you reckon?

Comment (georgegillams) [reply dangerfile.js]: Sure - I'm a fan of keeping it simple 😸

File: dangerfile.js [hunk 5]
```diff
@@ -30,27 +31,28 @@ const fileChanges = [...modifiedFiles, ...createdFiles];
...
+// Ensure new native components are extensible by consumers.
 const nativeComponentIntroduced = createdFiles.some(filePath => (
   filePath.match(/native\/packages\/react-native-bpk-component.+\/src\/.+\.js/)
 ));
 
 if (nativeComponentIntroduced) {
-  message('It looks like you are introducing a native component');
-  warn('Ensure the component style is extensible via `style`');
+  message('It looks like you are introducing a native component.');
+  warn('Ensure the component style is extensible via `style`.');
```
Comment (matteo-hertel) [inline dangerfile.js]: I feel like we could do some fancy checking on `propTypes` of the native component and break the build if style is not there, wdyt?

Comment (shaundon) [reply dangerfile.js]: We probably could, but I feel like that should be done in a further PR so as not to conflate this one

File: dangerfile.js [hunk 6]
```diff
@@ -71,3 +73,27 @@ const snapshotsModified = fileChanges.some(filePath => (
...
+// New files should include the Backpack license heading.
+const unlicensedFiles = createdFiles.filter((filePath) => {
+  if (filePath.match(/\.(js|css|scss|sh)/)) {
+    const fileContent = fs.readFileSync(filePath);
+    return !fileContent.includes('Licensed under the Apache License, Version 2.0 (the "License")');
+  }
+  return false;
+});
+if (unlicensedFiles.length > 0) {
+  fail(`These new files are do not include the license heading: ${unlicensedFiles.join(', ')}`);
```
Comment (matteo-hertel) [inline dangerfile.js]: what if we're adding a new file in e.g. new file in the dist folder in `bpk-svgs`?

Comment (matteo-hertel) [reply dangerfile.js]: oh merged already! I guess we'll find out 😄

Comment (shaundon) [reply dangerfile.js]: Hmm yeah good point. Maybe it should have just been a `warn` in the first place.

Comment (matteo-hertel) [reply dangerfile.js]: I'm adding some files in, if it's breaking something we'll fix it

Comment (shaundon) [reply dangerfile.js]: So I think we have two options here:

1) Change `fail` to `warn`.
2) Change Danger to disregard files inside `dist` folders.

Option 1 is my preference - I see it like the changelog where Danger will be right most of the time but sometimes we disregard the warning for good reason.

@matteo-hertel @matthewdavidson @georgegillams thoughts?

PS this is already merged so any changes will come in another one.

Comment (matthewdavidson) [reply dangerfile.js]: It should be impossible to add a new file without the license header. If there are directories of files that we need to ignore then that's fine (analogous to `.gitignore` / `.eslintignore`)

Comment (matthewdavidson) [reply dangerfile.js]: If we only have danger emitting warnings then eventually they'll become visual noise - humans will ignore them and eventually they'll cease to be of use.

Comment (georgegillams) [reply dangerfile.js]: Yeah I think we should disregard anything autogenerated (which I think is just inside `dist`??)

Comment (shaundon) [reply dangerfile.js]: https://github.com/Skyscanner/backpack/pull/144 sorts this out

**Pull request merged**